internal/servers/controller/internal/metrics/api.go

```diff
@@ -0,0 +1,285 @@
+// their respective functions ensures they remain singletons and allows
+// the code to enforce the appropriate labels are used.
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/boundary/globals"
+	"github.com/hashicorp/boundary/internal/gen/controller/api/services"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	grpcpb "google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+var (
+	pathRegex              map[*regexp.Regexp]string
+	expectedPathsToMethods map[string][]string
+)
+
+func init() {
+	expectedPathsToMethods = apiPathsAndMethods()
+	pathRegex = make(map[*regexp.Regexp]string)
+	for p := range expectedPathsToMethods {
+		pathRegex[buildRegexFromPath(p)] = p
+	}
+}
+
+// gatherPathInfo populates the provided map with the paths associated with
+// the provided HttpRule proto option as well as the path information contained
+// in any embedded AdditionalBindings in that HttpRule.
+func gatherPathInfo(rule *grpcpb.HttpRule, paths map[string][]string) {
+	switch r := rule.GetPattern().(type) {
+	case *grpcpb.HttpRule_Get:
+		paths[r.Get] = append(paths[r.Get], http.MethodGet)
+	case *grpcpb.HttpRule_Post:
+		paths[r.Post] = append(paths[r.Post], http.MethodPost)
+	case *grpcpb.HttpRule_Patch:
+		paths[r.Patch] = append(paths[r.Patch], http.MethodPatch)
+	case *grpcpb.HttpRule_Put:
+		paths[r.Put] = append(paths[r.Put], http.MethodPut)
+	case *grpcpb.HttpRule_Delete:
+		paths[r.Delete] = append(paths[r.Delete], http.MethodDelete)
+	default:
+		panic("unknown rule of ")
+	}
+	for _, additional := range rule.AdditionalBindings {
+		gatherPathInfo(additional, paths)
+	}
+}
+
+func gatherServicePathsAndMethods(fd protoreflect.FileDescriptor, paths map[string][]string) error {
+	for j := 0; j < fd.Services().Len(); j++ {
+		sd := fd.Services().Get(j)
+		for i := 0; i < sd.Methods().Len(); i++ {
+			r := sd.Methods().Get(i)
+			opts := r.Options().(*descriptorpb.MethodOptions)
+			httpRule := proto.GetExtension(opts, grpcpb.E_Http).(*grpcpb.HttpRule)
+			if proto.Equal(httpRule, &grpcpb.HttpRule{}) || httpRule == nil {
+				return fmt.Errorf("empty or no http rule found on service method %q", r.FullName())
+			}
+			gatherPathInfo(httpRule, paths)
+		}
+	}
+	return nil
+}
+
+// apiPathsAndMethods provides the pathing information for all services
+// registered to the controller API.  This does not include any paths that are
+// defined outside of the protobufs such as anything using the OPTION method
+// or any other paths registered like the dev UI passthrough path.
+func apiPathsAndMethods() map[string][]string {
+	// TODO: Auto generate the list of file descriptors so new services created
+	//  will automatically be covered by the metrics.
```

Comment:
I think this should be easy to do in this PR, you should be able to use [protoregistry.GlobalFiles](https://pkg.go.dev/google.golang.org/protobuf/reflect/protoregistry#Files), something like:

```go
paths := make(map[string][]string)
protoregistry.GlobalFiles.RangeFiles(func(f protoreflect.FileDescriptor) bool {
  if err := gatherServicePathsAndMethods(f, paths); err != nil {
    panic(err)
  }
  return true
})